src/base/SolutionArray.cpp

```diff
@@ -404,37 +815,200 @@ AnyMap SolutionArray::readHeader(const AnyMap& root, const std::string& id)
     return header;
 }
 
-AnyMap SolutionArray::restore(const std::string& fname, const std::string& id)
+AnyMap SolutionArray::restore(const string& fname,
+                              const string& id, const string& sub)
 {
     size_t dot = fname.find_last_of(".");
-    std::string extension = (dot != npos) ? toLowerCopy(fname.substr(dot + 1)) : "";
+    string extension = (dot != npos) ? toLowerCopy(fname.substr(dot + 1)) : "";
+    AnyMap header;
     if (extension == "h5" || extension == "hdf"  || extension == "hdf5") {
-        readEntry(fname, id);
-        return readHeader(fname, id);
-    }
-    if (extension == "yaml" || extension == "yml") {
+        readEntry(fname, id, sub);
+        header = readHeader(fname, id);
+    } else if (extension == "yaml" || extension == "yml") {
         const AnyMap& root = AnyMap::fromYamlFile(fname);
-        readEntry(root, id);
-        return readHeader(root, id);
+        readEntry(root, id, sub);
+        header = readHeader(root, id);
+    } else {
+        throw CanteraError("SolutionArray::restore",
+            "Unknown file extension '{}'; supported extensions include "
+            "'h5'/'hdf'/'hdf5' and 'yml'/'yaml'.", extension);
+    }
+    return header;
+}
+
+void SolutionArray::_initExtra(const string& name, const AnyValue& value)
+{
+    if (!m_extra->count(name)) {
+        throw CanteraError("SolutionArray::_initExtra",
+            "Component '{}' does not exists.", name);
```

Comment:
This appears a few times:
```suggestion
            "Component '{}' does not exist.", name);
```